pkg/register/authorization.go

```diff
@@ -27,51 +26,58 @@ type recaptchaResponse struct {
 	ErrorCodes  []string `json:"error-codes"`
 }
 
-// MakeHTTPRecaptchaValidationMW retrieves the recaptcha code and checks its validity
-func MakeHTTPRecaptchaValidationMW(recaptchaURL string, recaptchaSecret string, logger log.Logger) func(http.Handler) http.Handler {
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			var recaptchaToken = req.Header.Get("Authorization")
-			var ctx = context.TODO()
-
-			if recaptchaToken == "" {
-				logger.Info(ctx, "Authorization Error", "Missing Authorization header")
-				httpErrorHandler(ctx, http.StatusForbidden, errors.New(errorhandler.MsgErrMissingParam+"."+errorhandler.AuthHeader), w)
-				return
-			}
-
-			if match, _ := regexp.MatchString(regexpRecaptchaToken, recaptchaToken); !match {
-				logger.Info(ctx, "Authorization Error", "Invalid recaptcha token")
-				httpErrorHandler(ctx, http.StatusForbidden, errors.New(errorhandler.MsgErrMissingParam+"."+errorhandler.BasicToken), w)
-				return
-			}
-
-			if !checkRecaptcha(ctx, recaptchaURL, recaptchaSecret, recaptchaToken, logger) {
-				httpErrorHandler(ctx, http.StatusForbidden, errors.New(errorhandler.MsgErrInvalidParam+"."+errorhandler.Token), w)
-				return
-			}
-
-			next.ServeHTTP(w, req)
-		})
+type authorizationComponentMW struct {
+	recaptchaURL    string
+	recaptchaSecret string
+	logger          log.Logger
+	next            Component
+}
+
+// MakeAuthorizationRegisterComponentMW checks authorization and return an error if the action is not allowed.
+func MakeAuthorizationRegisterComponentMW(recaptchaURL, recaptchaSecret string, logger log.Logger) func(Component) Component {
+	return func(next Component) Component {
+		return &authorizationComponentMW{
+			recaptchaURL:    recaptchaURL,
+			recaptchaSecret: recaptchaSecret,
+			logger:          logger,
+			next:            next,
+		}
 	}
 }
 
-func httpErrorHandler(_ context.Context, statusCode int, err error, w http.ResponseWriter) {
-	w.WriteHeader(statusCode)
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+func (c *authorizationComponentMW) RegisterUser(ctx context.Context, socialRealmName, clientRealmName string, user apiregister.UserRepresentation) (string, error) {
```

Comment:
From my point of view, the recaptcha should be check at a higher level. I was more convinced by the middleware mechanism

Reply:
Rolled back